File: tree.java
import java.io.*;
import java.util.*;

class tree {
  void tree(String depth,File dir) {
    File[] files = dir.listFiles();
    int cnt = 0; 

    Arrays.sort(files);

    if (files == null) {
      return;
    }
    for(File file: files) {
      cnt++;
      if (file.isDirectory()) {
        if (files.length == cnt) {
          System.out.printf("%s`-- %s\n",depth,file.getName());
          tree(depth + "    ",new File(file.getAbsolutePath())); 
        }
        else {
          System.out.printf("%s|-- %s\n",depth,file.getName());
          tree(depth + "|   ",new File(file.getAbsolutePath())); 
        }
      }
      else {
        if (files.length == cnt) {
          System.out.printf("%s`-- %s\n",depth,file.getName());
        }
        else {
          System.out.printf("%s|-- %s\n",depth,file.getName());
        }
      }
    }
  }

  public static void main(String args[]) {
    String root_dir = ".";

    if (args.length != 0) {
      root_dir = args[0];
    }
    System.out.printf("%s\n",root_dir);

    tree obj = new tree();
    obj.tree("",new File(root_dir));
  }
}

File: tree.py
#! /usr/bin/python

import os
import sys

def tree(depth,dir):
  files = sorted(os.listdir(dir))
  cnt = 0

  for file in files:
    fullpath =  dir + "/" + file
    cnt = cnt + 1

    if os.path.isdir(fullpath):
      if len(files) == cnt:
        print "%s`-- %s"%(depth,file)
        tree(depth + "    " ,fullpath)
      else:
        print "%s|-- %s"%(depth,file)
        tree(depth + "|   " ,fullpath)
    else:
      if len(files) == cnt:
        print "%s`-- %s"%(depth,file)
      else:
        print "%s|-- %s"%(depth,file)

root_dir = "."
if len(sys.argv) > 1:
  root_dir = sys.argv[1]

print root_dir
tree("",root_dir)
